File: camelcase-matching.cpp
//Time - O(NxK) k - max (query, pattern)
//Space - O(N)
class Solution {
public:
    vector<bool> camelMatch(vector<string>& queries, string pattern) {
        vector<bool> result;
        
        for(auto query:queries){
            bool flag = false;
            int i = 0;
            for(int j = 0;j<query.size();j++){
                char c = query[j];
                if(i<pattern.size() && c == pattern[i]){
                    i++;
                    if(i == pattern.size()){
                        flag = true;
                    }
                }else if(isupper(c)){
                    flag = false;
                    break;
                }
            }
            result.push_back(flag);
        }
       
        return result;
        
    }
};

File: Camelcase Matching.py
class Solution:
    def match(self, query, pattern):
        pidx = 0
        for i in range(len(query)):
            if query[i].isupper():
                if query[i]==pattern[pidx]:
                    pidx+=1
                else:
                    return False
            else:
                if query[i]==pattern[pidx]:
                    pidx+=1  
            if pidx==len(pattern):
                for j in range(i+1, len(query)):
                    if query[j].isupper():
                        return False
                return True

            
    def camelMatch(self, queries: List[str], pattern: str) -> List[bool]:
        ans = [False for _ in range(len(queries))]
        for i in range(len(queries)):
            if self.match(queries[i], pattern):
                ans[i]=True
        return ans
